src/Microsoft.ML.Tokenizers/Model/SentencePieceBpe.cs

```diff
@@ -0,0 +1,949 @@
+        private const int ByteCodeToIdOffset = 3; // byte code is mapped to the to the Ids starting from 4.
+        const int OneUtf8ByteMaxId = 0x7F + ByteCodeToIdOffset; // 0x7F is the maximum value of the one byte UTF-8 character.
+
+        private readonly Dictionary<StringSpanOrdinalKey, (int Id, float Score, byte Type)> _vocab = new();
+        private readonly Dictionary<int, string> _vocabReverse = new();
+        private Dictionary<string, int>? _publicVocab;
+        private readonly int _maxByteId;
+
+        internal SentencePieceBpe(ModelProto modelProto, bool addBos, bool addEos) :
+                this(modelProto is null ? throw new ArgumentNullException(nameof(modelProto)) : modelProto)
+        {
+            AddBeginningOfSentence = addBos;
+            AddEndOfSentence = addEos;
+        }
+
+        private SentencePieceBpe(ModelProto modelProto)
+        {
+            for (int i = 0; i < modelProto.Pieces.Count; i++)
+            {
+                var piece = modelProto.Pieces[i];
+                _vocab.Add(new StringSpanOrdinalKey(piece.Piece), (i, piece.Score, (byte)piece.Type));
+                _vocabReverse.Add(i, piece.Piece);
+
+                if (piece.Type == ModelProto.Types.SentencePiece.Types.Type.Byte)
+                {
+                    _maxByteId = i;
+                }
+            }
+
+            BeginningOfSentenceToken = modelProto.TrainerSpec.BosPiece ?? "<s>";
+            BeginningOfSentenceId = modelProto.TrainerSpec.BosId <= 0 ? 1 : modelProto.TrainerSpec.BosId;
+            EndOfSentenceToken = modelProto.TrainerSpec.EosPiece ?? "</s>";
+            EndOfSentenceId = modelProto.TrainerSpec.EosId <= 0 ? 1 : modelProto.TrainerSpec.EosId;
+            UnknownToken = modelProto.TrainerSpec.UnkPiece ?? "<unk>";
+            UnknownId = modelProto.TrainerSpec.UnkId < 0 ? 0 : modelProto.TrainerSpec.UnkId;
+
+            AddDummyPrefix = modelProto.NormalizerSpec.AddDummyPrefix;
+            EscapeWhiteSpaces = modelProto.NormalizerSpec.EscapeWhitespaces;
+            TreatWhitespaceAsSuffix = modelProto.TrainerSpec.TreatWhitespaceAsSuffix;
+            ByteFallback = modelProto.TrainerSpec.ByteFallback;
+        }
+
+        /// <summary>
+        /// Create a SentencePieceBpe tokenizer from the given model stream. The model stream should contain the SentencePiece Bpe model according to
+        /// https://github.com/google/sentencepiece/blob/master/src/sentencepiece_model.proto specification.
+        /// </summary>
+        /// <param name="modelStream">The stream containing the SentencePiece Bpe model.</param>
+        /// <param name="addBeginOfSentence">Indicate emitting the beginning of sentence token during the encoding.</param>
+        /// <param name="addEndOfSentence">Indicate emitting the end of sentence token during the encoding.</param>
+        public static Tokenizer CreateLlamaTokenizer(
```

Comment:
@stephentoub I guess I need to provide async version of this method too. I'll try to provide it before merging this PR.

Reply:
We discussed this matter offline and have decided not to provide an asynchronous version for now, as ModelProto.Parser.ParseFrom does not offer an asynchronous version. However, we may consider exposing the asynchronous version of CreateLlamaTokenizer if circumstances change in the future or if there is a demand for it.